File: data/llama3_8B_ins_results/readable_format_HPEC/fortran/file165.f90
subroutine sparse( a, colidx, rowstr, n, nz, nonzer, arow, acol,  &
&                   aelt, firstrow, lastrow,  &
&                   v, iv, nzloc, rcond, shift )

use tinfo

implicit           none

integer            colidx(*), iv(*)
integer            firstrow, lastrow
integer            n, nonzer, arow(*), acol(nonzer+1,*)
integer(kz)        nz, rowstr(*)
double precision   a(*), aelt(nonzer+1,*), v(*), rcond, shift

integer            nzloc(n), nrows


integer            i, j, jcol
integer(kz)        j1, j2, nza, k, kk, nzrow
double precision   xi, size, scale, ratio, va

nrows = lastrow - firstrow + 1
j1 = ilow + 1
j2 = ihigh + 1

do j = j1, j2
rowstr(j) = 0
enddo

do i = 1, n
do nza = 1, arow(i)
j = acol(nza, i)
if (j.ge.ilow .and. j.le.ihigh) then
j = j + 1
rowstr(j) = rowstr(j) + arow(i)
endif
end do
end do

if (myid .eq. 0) then
rowstr(1) = 1
j1 = 1
endif
do j = j1+1, j2
rowstr(j) = rowstr(j) + rowstr(j-1)
enddo
if (myid .lt. num_threads) last_n(myid) = rowstr(j2)

nzrow = 0
if (myid .lt. num_threads) then
do i = 0, myid-1
nzrow = nzrow + last_n(i)
end do
endif
if (nzrow .gt. 0) then
do j = j1, j2
rowstr(j) = rowstr(j) + nzrow
enddo
endif
nza = rowstr(nrows+1) - 1


if (nza .gt. nz) then
write(*,*) 'Space for matrix elements exceeded in sparse'
write(*,*) 'nza, nzmax = ',nza, nz
stop
endif


do j = ilow, ihigh
do k = rowstr(j), rowstr(j+1)-1
v(k) = 0.d0
iv(k) = 0
enddo
nzloc(j) = 0
enddo


size = 1.0D0
ratio = rcond ** (1.0D0 / dfloat(n))

do i = 1, n
do nza = 1, arow(i)
j = acol(nza, i)

if (j .lt. ilow .or. j .gt. ihigh) goto 60

scale = size * aelt(nza, i)
do nzrow = 1, arow(i)
jcol = acol(nzrow, i)
va = aelt(nzrow, i) * scale

if (jcol .eq. j .and. j .eq. i) then
va = va + rcond - shift
endif

do k = rowstr(j), rowstr(j+1)-1
if (iv(k) .gt. jcol) then
do kk = rowstr(j+1)-2, k, -1
if (iv(kk) .gt. 0) then
v(kk+1)  = v(kk)
iv(kk+1) = iv(kk)
endif
enddo
iv(k) = jcol
v(k)  = 0.d0
goto 40
else if (iv(k) .eq. 0) then
iv(k) = jcol
goto 40
else if (iv(k) .eq. jcol) then
nzloc(j) = nzloc(j) + 1
goto 40
endif
enddo
print *,'internal error in sparse: i=',i
stop
40          continue
v(k) = v(k) + va
enddo
60       continue
enddo
size = size * ratio
enddo


do j = ilow+1, ihigh
nzloc(j) = nzloc(j) + nzloc(j-1)
enddo
if (myid .lt. num_threads) last_n(myid) = nzloc(ihigh)

nzrow = 0
if (myid .lt. num_threads) then
do i = 0, myid-1
nzrow = nzrow + last_n(i)
end do
endif
if (nzrow .gt. 0) then
do j = ilow, ihigh
nzloc(j) = nzloc(j) + nzrow
enddo
endif

do j = 1, nrows
if (j .gt. 1) then
j1 = rowstr(j) - nzloc(j-1)
else
j1 = 1
endif
j2 = rowstr(j+1) - nzloc(j) - 1
nza = rowstr(j)
do k = j1, j2
a(k) = v(nza)
colidx(k) = iv(nza)
nza = nza + 1
enddo
enddo
do j = 2, nrows+1
rowstr(j) = rowstr(j) - nzloc(j-1)
enddo
nza = rowstr(nrows+1) - 1


return
11000   format ( //,'final nonzero count in sparse ',  &
&            /,'number of nonzeros       = ', i16 )
end

File: data/llama3_8B_ins_results/readable_format_HPEC/cpp/file165.cpp
#include <iostream>
#include <vector>

using namespace std;

void sparse(double* a, int* colidx, int* rowstr, int n, int nz, double* aelt, int* firstrow, int* lastrow, double* v, int* iv, int* nzloc, double rcond, double shift) {
  int i, j, jcol, k, kk, nzrow;
  double xi, size, scale, ratio, va;

  int nrows = *lastrow - *firstrow + 1;
  int j1 = 1;
  int j2 = *ihigh + 1;

  for (j = j1; j <= j2; j++) {
    rowstr[j] = 0;
  }

  for (i = 1; i <= n; i++) {
    int nza = 1;
    for (nza = 1; nza <= *arow[i]; nza++) {
      j = *acol[nza][i];
      if (j >= *ilow && j <= *ihigh) {
        j++;
        rowstr[j] = rowstr[j] + *arow[i];
      }
    }
  }

  if (*myid == 0) {
    rowstr[1] = 1;
    j1 = 1;
  }
  for (j = j1 + 1; j <= j2; j++) {
    rowstr[j] = rowstr[j] + rowstr[j - 1];
  }
  if (*myid < *num_threads) {
    *last_n[*myid] = rowstr[j2];
  }

  int nzrow = 0;
  if (*myid < *num_threads) {
    for (i = 0; i < *myid; i++) {
      nzrow += *last_n[i];
    }
  }
  if (nzrow > 0) {
    for (j = j1; j <= j2; j++) {
      rowstr[j] = rowstr[j] + nzrow;
    }
  }

  int nza = rowstr[nrows + 1] - 1;

  if (nza > nz) {
    cout << "Space for matrix elements exceeded in sparse" << endl;
    cout << "nza, nzmax = " << nza << ", " << nz << endl;
    exit(1);
  }

  for (j = *ilow; j <= *ihigh; j++) {
    for (k = rowstr[j]; k < rowstr[j + 1] - 1; k++) {
      v[k] = 0.0;
      iv[k] = 0;
    }
    nzloc[j] = 0;
  }

  size = 1.0;
  ratio = pow(*rcond, 1.0 / *n);

  for (i = 1; i <= n; i++) {
    int nza = 1;
    for (nza = 1; nza <= *arow[i]; nza++) {
      j = *acol[nza][i];
      if (j < *ilow || j > *ihigh) continue;
      scale = size * *aelt[nza][i];
      for (nzrow = 1; nzrow <= *arow[i]; nzrow++) {
        jcol = *acol[nzrow][i];
        va = *aelt[nzrow][i] * scale;
        if (jcol == j && j == i) {
          va += *rcond - *shift;
        }
        for (k = rowstr[j]; k < rowstr[j + 1] - 1; k++) {
          if (iv[k] > jcol) {
            for (kk = rowstr[j + 1] - 2; kk >= k; kk--) {
              if (iv[kk] > 0) {
                v[kk + 1] = v[kk];
                iv[kk + 1] = iv[kk];
              }
            }
            iv[k] = jcol;
            v[k] = 0.0;
            break;
          } else if (iv[k] == 0) {
            iv[k] = jcol;
            break;
          } else if (iv[k] == jcol) {
            nzloc[j]++;
            break;
          }
        }
        v[k] += va;
      }
    }
  }

  size *= ratio;

  for (j = *ilow + 1; j <= *ihigh; j++) {
    nzloc[j] += nzloc[j - 1];
  }
  if (*myid < *num_threads) {
    *last_n[*myid] = nzloc[*ihigh];
  }

  nzrow = 0;
  if (*myid < *num_threads) {
    for (i = 0; i < *myid; i++) {
      nzrow += *last_n[i];
    }
  }
  if (nzrow > 0) {
    for (j = *ilow; j <= *ihigh; j++) {
      nzloc[j] += nzrow;
    }
  }

  for (j = 1; j <= nrows; j++) {
    if (j > 1) {
      j1 = rowstr[j] - nzloc[j - 1];
    } else {
      j1 = 1;
    }
    j2 = rowstr[j + 1] - nzloc[j] - 1;
    nza = rowstr[j];
    for (k = j1; k <= j2; k++) {
      a[k] = v[nza];
      colidx[k] = iv[nza];
      nza++;
    }
  }

  for (j = 2; j <= nrows + 1; j++) {
    rowstr[j] = rowstr[j] - nzloc[j - 1];
  }
  nza = rowstr[nrows + 1] - 1;

  return;
}

int main() {
  // Initialize variables
  //...

  // Call the sparse function
  sparse(a, colidx, rowstr, n, nz, aelt, firstrow, lastrow, v, iv, nzloc, rcond, shift);

  //...
  return 0;
}
